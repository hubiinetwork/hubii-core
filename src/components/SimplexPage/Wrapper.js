import styled from 'styled-components';

export default styled.div`
  width: 100%;
  height: 85vh;
`;
