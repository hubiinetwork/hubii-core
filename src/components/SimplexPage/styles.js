import styled from 'styled-components';
import { Spin } from 'antd';

export const Wrapper = styled.div`
  display: flex;
  height: 100%;
  align-items: center;
  justify-content: center;
  flex-direction: column;
  flex: 1;
`;

export const StyledSpin = styled(Spin)`
  margin-top: 10rem;
  color: white;
`;
