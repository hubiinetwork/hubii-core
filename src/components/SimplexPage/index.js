/**
*
* SimplexPage
*
*/

import React from 'react';
import PropTypes from 'prop-types';
import { isValidAddress } from 'ethereumjs-util';
import { Wrapper, StyledSpin } from './styles';
import ScrollableContentWrapper from '../ui/ScrollableContentWrapper';


class SimplexPage extends React.PureComponent { // eslint-disable-line react/prefer-stateless-function
  constructor(props) {
    super(props);
    this.state = { loading: false, error: false, timesStoppedLoading: 0 };
    this.startLoad = this.startLoad.bind(this);
    this.stopLoad = this.stopLoad.bind(this);
  }

  componentDidMount() {
    const webview = document.querySelector('webview');
    if (webview) {
      webview.addEventListener('did-start-loading', this.startLoad);
      webview.addEventListener('did-stop-loading', this.stopLoad);
    } else {
      this.setState({ error: true }); // eslint-disable-line
    }
  }

  // will show loading for landing page load and simplex checkout
  // load. stop checking after that, since it'll flash the loading
  // screen a few times during the simplex checkout process
  startLoad() {
    if (this.state.timesStoppedLoading < 2) {
      this.setState({ loading: true, timesStoppedLoading: this.state.timesStoppedLoading += 1 });
    }
  }

  stopLoad() {
    if (this.state.loading) {
      this.setState({ loading: false });
    }
  }

  render() {
    const { match } = this.props;
    const { loading, error } = this.state;
    const address = match.path.substring(8, 50);
    if (!isValidAddress(address) || error) {
      return (
        <p style={{ color: 'white' }}>Unexpected error occured, please try again later</p>
      );
    }
    return (
      <ScrollableContentWrapper>
        <Wrapper>
          {
            loading &&
            <StyledSpin tip="Loading..." size="large" />
          }
          <webview
            style={loading ? { visibility: 'hidden' } : { width: '100%', height: '80vh' }}
            autosize="on"
            src={'https://simplex.dev.hubii.net/?crypto=ETH'}
          ></webview>
        </Wrapper>
      </ScrollableContentWrapper>
    );
  }
}

SimplexPage.propTypes = {
  match: PropTypes.object.isRequired,
};

export default SimplexPage;
