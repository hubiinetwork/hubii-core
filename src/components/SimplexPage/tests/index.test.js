import React from 'react';
import { shallow } from 'enzyme';

import SimplexPage from '../index';

describe('<SimplexPage />', () => {
  const props = {
    match: {
      path: '/wallet/0x910c4BA923B2243dc13e00A066eEfb8ffd905EB0/buyeth',
    },
  };
  it('should render correctly when no errors', () => {
    const wrapper = shallow(<SimplexPage {...props} />);
    wrapper.setState({ error: false });
    expect(wrapper).toMatchSnapshot();
  });

  it('should render correctly on error', () => {
    const wrapper = shallow(<SimplexPage {...props} />);
    wrapper.setState({ error: true });
    expect(wrapper).toMatchSnapshot();
  });

  it('should render correctly when couldn\'t get address from url', () => {
    const wrapper = shallow(<SimplexPage match={{ path: '1231' }} />);
    wrapper.setState({ error: false });
    expect(wrapper).toMatchSnapshot();
  });

  it('should render correctly when loading', () => {
    const wrapper = shallow(<SimplexPage {...props} />);
    wrapper.setState({ error: false, loading: true });
    expect(wrapper).toMatchSnapshot();
  });
});
