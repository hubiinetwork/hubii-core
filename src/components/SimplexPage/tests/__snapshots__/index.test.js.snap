// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SimplexPage /> should render correctly on error 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <SimplexPage
    match={
        Object {
            "path": "/wallet/0x910c4BA923B2243dc13e00A066eEfb8ffd905EB0/buyeth",
          }
    }
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": "Unexpected error occured, please try again later",
      "style": Object {
        "color": "white",
      },
    },
    "ref": null,
    "rendered": "Unexpected error occured, please try again later",
    "type": "p",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": "Unexpected error occured, please try again later",
        "style": Object {
          "color": "white",
        },
      },
      "ref": null,
      "rendered": "Unexpected error occured, please try again later",
      "type": "p",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;

exports[`<SimplexPage /> should render correctly when couldn't get address from url 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <SimplexPage
    match={
        Object {
            "path": "1231",
          }
    }
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": "Unexpected error occured, please try again later",
      "style": Object {
        "color": "white",
      },
    },
    "ref": null,
    "rendered": "Unexpected error occured, please try again later",
    "type": "p",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": "Unexpected error occured, please try again later",
        "style": Object {
          "color": "white",
        },
      },
      "ref": null,
      "rendered": "Unexpected error occured, please try again later",
      "type": "p",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;

exports[`<SimplexPage /> should render correctly when loading 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <SimplexPage
    match={
        Object {
            "path": "/wallet/0x910c4BA923B2243dc13e00A066eEfb8ffd905EB0/buyeth",
          }
    }
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "children": Array [
        <styles__StyledSpin
          size="large"
          tip="Loading..."
/>,
        <webview
          autosize="on"
          src="https://simplex.dev.hubii.net/?crypto=ETH"
          style={
                    Object {
                              "visibility": "hidden",
                            }
          }
/>,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "size": "large",
          "tip": "Loading...",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "autosize": "on",
          "src": "https://simplex.dev.hubii.net/?crypto=ETH",
          "style": Object {
            "visibility": "hidden",
          },
        },
        "ref": null,
        "rendered": null,
        "type": "webview",
      },
    ],
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "children": Array [
          <styles__StyledSpin
            size="large"
            tip="Loading..."
/>,
          <webview
            autosize="on"
            src="https://simplex.dev.hubii.net/?crypto=ETH"
            style={
                        Object {
                                    "visibility": "hidden",
                                  }
            }
/>,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "size": "large",
            "tip": "Loading...",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "autosize": "on",
            "src": "https://simplex.dev.hubii.net/?crypto=ETH",
            "style": Object {
              "visibility": "hidden",
            },
          },
          "ref": null,
          "rendered": null,
          "type": "webview",
        },
      ],
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;

exports[`<SimplexPage /> should render correctly when no errors 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <SimplexPage
    match={
        Object {
            "path": "/wallet/0x910c4BA923B2243dc13e00A066eEfb8ffd905EB0/buyeth",
          }
    }
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "children": Array [
        false,
        <webview
          autosize="on"
          src="https://simplex.dev.hubii.net/?crypto=ETH"
          style={
                    Object {
                              "height": "80vh",
                              "width": "100%",
                            }
          }
/>,
      ],
    },
    "ref": null,
    "rendered": Array [
      false,
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "autosize": "on",
          "src": "https://simplex.dev.hubii.net/?crypto=ETH",
          "style": Object {
            "height": "80vh",
            "width": "100%",
          },
        },
        "ref": null,
        "rendered": null,
        "type": "webview",
      },
    ],
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "children": Array [
          false,
          <webview
            autosize="on"
            src="https://simplex.dev.hubii.net/?crypto=ETH"
            style={
                        Object {
                                    "height": "80vh",
                                    "width": "100%",
                                  }
            }
/>,
        ],
      },
      "ref": null,
      "rendered": Array [
        false,
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "autosize": "on",
            "src": "https://simplex.dev.hubii.net/?crypto=ETH",
            "style": Object {
              "height": "80vh",
              "width": "100%",
            },
          },
          "ref": null,
          "rendered": null,
          "type": "webview",
        },
      ],
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
